readme.md
# Livewire - Notifier.

[![Latest Stable Version](https://poser.pugx.org/slakbal/tall-notifier/v)](//packagist.org/packages/slakbal/tall-notifier) [![run-tests](https://github.com/slakbal/tall-notifier/actions/workflows/run-tests.yml/badge.svg)](https://github.com/slakbal/tall-notifier/actions/workflows/run-tests.yml) [![Total Downloads](https://poser.pugx.org/slakbal/tall-notifier/downloads)](//packagist.org/packages/slakbal/tall-notifier) [![License](https://poser.pugx.org/slakbal/tall-notifier/license)](//packagist.org/packages/phpunit/phpunit)


Livewire Notifier is a simple notifications system with zero dependencies above [TALL-stack](https://tallstack.dev/) (Tailwind CSS, Alpine.JS, Laravel and Livewire).
<p align="center">
    <img src="https://github.com/slakbal/tall-notifier/blob/main/screenshots/tall-notifier.jpg?raw=true" alt="Livewire Notifier">
</p>

## Requirements

Make sure that [Livewire](https://laravel-livewire.com/) and [Alpine.JS](https://github.com/alpinejs/alpine) are installed properly.
The easiest way to do it is to install [Laravel Jetstream](https://jetstream.laravel.com) with **Livewire stack** (post-install command `php artisan jetstream:install livewire`).

**Alpine.JS** must be imported in `resources/js/app.js` package:
```javascript
import 'alpinejs'
```
**Livewire** scripts and styles tags must be present in the layout file:
```html
…
<head>
…
    <livewire:styles/>
…
</head>
<body>
…
    <livewire:scripts/>
</body>
</html>
```
## Installation

Via Composer

```bash
$ composer require slakbal/tall-notifier
```

Proceed with installation process:

```bash
$ php artisan tall-notifier:install
```

Afterwards the package config can be found at `config/tall-notifier.php` and views at `resources/views/vendor/tall-notifier/`.

It's required because of Tailwind config is needed to be extended with new purge paths.

Otherwise you won't see messages styled properly.

<p align="center">
    <img src="https://github.com/slakbal/tall-notifier/blob/main/screenshots/tall-notifier-basic.jpg?raw=true" alt="Livewire Notifier">
</p>

## Usage

Put Livewire-component `<livewire:notifier/>` into the app layout.
Make sure to insert it into correct context because it may be positioned absolutely.

Now you can use it from frontend and backend both.

## Message options

Message added at backend (from any Livewire component) must have type of `array`.
Message added at frontend (from JavaScript) must have type of `object`.

``` php
$message = [
        'text' => __('Post is saved!'),
        'title' => '', // Optional
        'type' => 'success', // Optional. By default: success | optional: error (or fail), warning (or warn), info
        // you can find all types options in the config file
        'icon' => '', // Optional. It replaces the default type icon. Can be pure html or svg code

        // Attention! The following options override ones from the config file

        'duration' => 5000, // Optional. The time of message to be shown. To show infinitely set to 0
        'msgClass' =>  'bg-gradient-to-r from-green-200 to-green-300', // Optional. Tailwind class for message container
        'progressClass' =>  'bg-green-500', // Optional. Tailwind class for progress bar. If null progress bar won't be shown
        'closable' => false, // Optional. True by default. Whether message is closable by click on message or Esc key press on window
    ]
```
``` javascript
let message = {
    text: 'Post is saved!'
}
```

### Backend

**Livewire event**

From any Livewire component push `emit` trigger to render new message.

``` php
public function save(){
    ...
    $this->emitTo('notifier','message',['text'=>__('The post is saved!')]);
}
```

**Session flash variable**
``` php
public function save(){
    ...
    session()->flash('notifier',['text'=>__('The post is saved!')]);
    return $this->redirect(route('posts'));
}
```

### Frontend
Add new message from javascript:
``` javascript
Livewire.emitTo('notifier','message',{text:'The post is saved!'})
```

### Example
Try to put the following code (with Laravel Jetstream w/Livewire stack installed):
```html
<div class="flex flex-row space-x-4">
    <x-jet-button x-data="{}" @click="Livewire.emitTo('notifier','message',{text:'Success',type:'success'})">Success</x-jet-button>
    <x-jet-button x-data="{}" @click="Livewire.emitTo('notifier','message',{text:'Error',type:'error'})">Error</x-jet-button>
    <x-jet-button x-data="{}" @click="Livewire.emitTo('notifier','message',{text:'Warning',type:'warning'})">Warning</x-jet-button>
    <x-jet-button x-data="{}" @click="Livewire.emitTo('notifier','message',{text:'Info',type:'info'})">Info</x-jet-button>
    <x-jet-button x-data="{}" @click="Livewire.emitTo('notifier','message',{text:'Default',type:'default'})">Default</x-jet-button>
</div>
```

## Config file

After `php artisan tall-notifier:install` command is fired, config file will be published to `config/tall-notifier.php`. There you can change some value for customization.
``` php
<?php
return [
    // Messages container positioning
    'positionClass' => 'absolute top-0 right-0',
    // Default message Tailwind stylinh
    'defaultMsgClass' => 'w-80 rounded-xl shadow-xl',
    // Time of each message to be shown by default
    'duration' => 2200,
    // Messages types
    'types' => [
        'success' => [
            // 'msgClass'=>'bg-green-200',
            'msgClass'=>'bg-gradient-to-r from-green-200 to-green-300',
            'progressClass' => 'bg-green-500',
            // View for icon
            'icon' => 'tall-notifier::icons.success',
        ],
        'error' => [
            // 'msgClass'=>'bg-red-200',
            'msgClass'=>'bg-gradient-to-r from-red-200 to-red-300',
            'progressClass' => 'bg-red-500',
            // View for icon
            'icon' => 'tall-notifier::icons.error',
        ],
        'fail' => [
            // 'msgClass'=>'bg-red-200',
            'msgClass'=>'bg-gradient-to-r from-red-200 to-red-300',
            'progressClass' => 'bg-red-500',
            // View for icon
            'icon' => 'tall-notifier::icons.error',
        ],
        'warning' => [
            // 'msgClass'=>'bg-yellow-200',
            'msgClass'=>'bg-gradient-to-r from-yellow-200 to-yellow-300',
            'progressClass' => 'bg-yellow-500',
            // View for icon
            'icon' => 'tall-notifier::icons.error',
        ],
        'warn' => [
            // 'msgClass'=>'bg-yellow-200',
            'msgClass'=>'bg-gradient-to-r from-yellow-200 to-yellow-300',
            'progressClass' => 'bg-yellow-500',
            // View for icon
            'icon' => 'tall-notifier::icons.error',
        ],
        'info' => [
            // 'msgClass'=>'bg-blue-200',
            'msgClass'=>'bg-gradient-to-r from-blue-200 to-blue-300',
            'progressClass' => 'bg-blue-500',
            // View for icon
            'icon' => 'tall-notifier::icons.info',
        ],
        'default' => [
            // 'msgClass'=>'bg-gray-200',
            'msgClass'=>'bg-gradient-to-r from-gray-200 to-gray-300',
            'progressClass' => 'bg-gray-700',
            // View for icon
            'icon' => 'tall-notifier::icons.info',
        ]
    ]
];
```

## Troubleshooting

Your messages don't get any styles? Please, run `tall-notifier:install` command to publish views. Therefore Laravel Mix compiler will find package Blade-views and will purge CSS properly.

## Updating

Simply update like all other packages with `composer update`. Be sure to run `tall-notifier:install` to updated published views.

## Change log

Please see the [changelog](changelog.md) for more information on what has changed recently.

## Testing

```bash
$ composer test
```

## Contributing

Please see [contributing.md](contributing.md) for details and a todolist.

## Security

If you discover any security related issues, please email author email instead of using the issue tracker.

## Credits

- [Dmitriy Belyaev](https://slakbal.com)

## License

Please see the [license file](license.md) for more information.


changelog.md
# Changelog

All notable changes to `LaravelTallNotifier` will be documented in this file.

## 1.0.3 2021-13-04
- **Changed** Readme update.
- **Changed** Update installation command.

## 1.0.2 2021-10-04
### Added
- Screenshots for README
- Changelog
### Changed
- README. Add requirements and installation instructions.
### Removed
- Delete unused boilerplate code.

## 1.0.1 2021-10-04
### Added
- Badges to readme
### Fixed
- Install command for proper tuning `tailwind.config.js` / add extra path for purging

tests/Pest.php
<?php

/*
|--------------------------------------------------------------------------
| Test Case
|--------------------------------------------------------------------------
|
| The closure you provide to your test functions is always bound to a specific PHPUnit test
| case class. By default, that class is "PHPUnit\Framework\TestCase". Of course, you may
| need to change it using the "uses()" function to bind a different classes or traits.
|
*/
uses(Slakbal\TallNotifier\Tests\TestCase::class)->in(__DIR__);
/*
|--------------------------------------------------------------------------
| Expectations
|--------------------------------------------------------------------------
|
| When you're writing tests, you often need to check that values meet certain conditions. The
| "expect()" function gives you access to a set of "expectations" methods that you can use
| to assert different things. Of course, you may extend the Expectation API at any time.
|
*/

// expect()->extend('toBeOne', function () {
//     return $this->toBe(1);
// });

/*
|--------------------------------------------------------------------------
| Functions
|--------------------------------------------------------------------------
|
| While Pest is very powerful out-of-the-box, you may have some testing code specific to your
| project that you don't want to repeat in every file. Here you can also expose helpers as
| global functions to help you to reduce the number of lines of code in your test files.
|
*/

// function something()
// {
//     // ..
// }


tests/TestCase.php
<?php

namespace Slakbal\TallNotifier\Tests;

use Illuminate\Database\Eloquent\Factories\Factory;
use Orchestra\Testbench\TestCase as Orchestra;
use Slakbal\TallNotifier\TallNotifierServiceProvider;
use Livewire\LivewireServiceProvider;

class TestCase extends Orchestra
{
    protected function setUp(): void
    {

        parent::setUp();
        // Factory::guessFactoryNamesUsing(
        //     fn (string $modelName) => 'Slakbal\\TallNotifier\\Database\\Factories\\'.class_basename($modelName).'Factory'
        // );
        // $this->viewsDirectory = __DIR__.'/views';
    }

    protected function getPackageProviders($app)
    {
        return [
            TallNotifierServiceProvider::class,
            LivewireServiceProvider::class
        ];
    }
    // protected function overrideApplicationProviders($app)
    // {
    //     return [
    //         'Livewire' => 'Livewire\Livewire',
    //     ];
    // }
    public function getEnvironmentSetUp($app)
    {
        $app['config']->set('database.default', 'sqlite');
        $app['config']->set('database.connections.sqlite', [
            'driver' => 'sqlite',
            'database' => ':memory:',
            'prefix' => '',
        ]);

        /*
        include_once __DIR__.'/../database/migrations/create_livewire_notifier_table.php.stub';
        (new \CreatePackageTable())->up();
        */
    }
}


tests/Unit/NotifierLivewireComponentTest.php
<?php
// use Livewire\Livewire;
use Livewire\Component;
use function Pest\Livewire\livewire;
use Slakbal\TallNotifier\Http\Livewire\Notifier;

beforeEach(function () {
    $this->message = [
        'text' => "Let's have fun!"
    ];
    $this->messageFromMessageBag = [
        'text' => "Let's have fun!",
        'title' => '',
        'icon' => '',
        'type' => 'success'
    ];
});
// it('accepts message argument', function () {
//     livewire(Notifier::class)->set('message',$this->message)->assertSet('message',$this->message);
// });
it('adds message to message bag received as an argument', function () {
    $livewire = livewire(Notifier::class, ['message' => $this->message]);
    $this->assertCount(1, $livewire->messages);
});
it('adds message to message bag by calling addMsg action', function () {
    $livewire = livewire(Notifier::class)->call('addMsg', $this->message);
    $this->assertCount(1, $livewire->messages);
    $this->assertEquals($this->messageFromMessageBag, $livewire->messages->pop());
});
it('grab message from session', function () {
    session()->flash('notifier', $this->message);
    $livewire = livewire(Notifier::class);
    $this->assertCount(1, $livewire->messages);
});
it('reacts on "notifier" event', function () {
    $this->message = [
        'text' => "Let's have fun!"
    ];
    $livewire = livewire(Notifier::class);
    $livewire->emit('message', $this->message);
    $this->assertCount(1, $livewire->messages);
});


src/Http/Livewire/Notifier.php
<?php

namespace Slakbal\TallNotifier\Http\Livewire;

use Livewire\Component;
use Illuminate\Support\Collection;

class Notifier extends Component
{
    public $messages;

    /**
     * Message from attributes bag
     *
     * @var String
     */
    public $message;
    /**
     * Duration of one message is shown
     *
     * @var integer
     */
    public $duration;
    public $listeners = [
        'message' => 'addMsg',
        'render' => 'render'
    ];
    /**
     * Catch session messages in the constructor
     *
     * @return void
     */
    public function mount()
    {
        $this->messages = collect($this->messages);
        $this->message && $this->addMsg($this->message);
        $this->reset('message');
        if ($msg = session('notifier')) {
            $this->addMsg($msg);
        }
        $this->duration = $this->duration ?? config('tall-notifier.duration');
    }
    /**
     * Add message to messages bag.
     *
     * @param mixed $options
     * @return void
     */
    public function addMsg($message)
    {
        if (!is_array($message) && $json = json_decode($message)) {
            $message = $json;
        } elseif (is_string($message)) {
            $message = ['text' => $message];
        }
        $message = array_merge([
            'text' => '',
            'title' => '',
            'icon' => '',
            'type' => 'success',
        ], $message);
        $this->messages->push($message);
    }

    public function render()
    {
        return view('tall-notifier::livewire.notifier');
    }
}


src/TallNotifierServiceProvider.php
<?php

namespace Slakbal\TallNotifier;

use Livewire\Livewire;
// use Illuminate\Support\Facades\Blade;
use Illuminate\Support\ServiceProvider;
use Slakbal\TallNotifier\Http\Livewire\Notifier;
use Slakbal\TallNotifier\Http\Livewire\NotifierMessage;
use Slakbal\TallNotifier\Console\Commands\TallNotifierInstall;

class TallNotifierServiceProvider extends ServiceProvider
{
    /**
     * Perform post-registration booting of services.
     *
     * @return void
     */
    public function boot(): void
    {
        // Publishing is only necessary when using the CLI.
        if ($this->app->runningInConsole()) {
            $this->bootForConsole();
        }
        Livewire::component('notifier', Notifier::class);
        Livewire::component('notifier-message', NotifierMessage::class);
        $this->loadViewsFrom(__DIR__ . '/../resources/views', 'tall-notifier');
    }

    /**
     * Register any package services.
     *
     * @return void
     */
    public function register(): void
    {
        $this->mergeConfigFrom(__DIR__ . '/../config/tall-notifier.php', 'tall-notifier');

        // Register the service the package provides.
        // $this->app->singleton('tall-notifier', function ($app) {
        //     return new TallNotifier;
        // });
    }

    // /**
    //  * Get the services provided by the provider.
    //  *
    //  * @return array
    //  */
    // public function provides()
    // {
    //     return ['tall-notifier'];
    // }

    /**
     * Console-specific booting.
     *
     * @return void
     */
    protected function bootForConsole(): void
    {
        // Publishing the configuration file.
        $this->publishes([
            __DIR__ . '/../config/tall-notifier.php' => config_path('tall-notifier.php'),
        ], 'tall-notifier.config');

        // Publishing the views.
        $this->publishes([
            __DIR__ . '/../resources/views' => base_path('resources/views/vendor/tall-notifier'),
        ], 'tall-notifier.views');

        // Registering package commands.
        $this->commands([TallNotifierInstall::class]);
    }
}


src/Http/Livewire/NotifierMessage.php
<?php

namespace Slakbal\TallNotifier\Http\Livewire;

use Livewire\Component;
use Illuminate\Support\Collection;

class NotifierMessage extends Component
{
    /**
     * Message from attributes bag
     *
     * @var String
     */
    public $message;
    /**
     * Time in ms which message is shown
     * 0 for unlimited time
     * @var integer
     */
    public $duration;

    public $msgClass;
    public $progressClass;
    /**
     * Whether message is closable by click
     *
     * @var boolean
     */
    public $closable = true;
    /**
     * Mount lifecycle action
     *
     * @return void
     */
    public function mount()
    {
        $this->duration = $this->duration ?? config('tall-notifier.duration');
        $this->closable = $this->closable ?? config('tall-notifier.closable');
        $this->msgClass = $this->msgClass ?? config('tall-notifier.types.' . ($this->message['type'] ?? 'default') . '.msgClass', config('tall-notifier.types.default.msgClass'));
        $this->progressClass = $this->progressClass ?? config('tall-notifier.types.' . ($this->message['type'] ?? 'default') . '.progressClass', config('tall-notifier.types.default.progressClass'));
    }

    public function render()
    {
        $this->message = array_merge([
            'text' => '',
            'title' => '',
            'icon' => '',
            'type' => 'success',
            'duration' => $this->duration,
            'msgClass' =>  $this->msgClass,
            'progressClass' =>  $this->progressClass,
            'closable' => $this->closable,
        ], $this->message);
        return view('tall-notifier::livewire.notifier-message');
    }
}


src/Console/Commands/LivewireNotifierInstall.php
<?php

namespace Slakbal\TallNotifier\Console\Commands;

use Illuminate\Console\Command;

class TallNotifierInstall extends Command
{
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'tall-notifier:install';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Livewire Notifier installation.';

    /**
     * Execute the console command.
     *
     * @return int
     */
    public function handle()
    {
        $this->table(['Livewire Notifier'], [
            ['Simple notifications system with zero dependencies above TALL-stack.'],
            ["\nVisit site:\nhttps://github.com/slakbal/tall-notifier"]
        ]);
        $this->info('😎 Installing Livewire Notifier…');
        if (!strpos(file_get_contents('./composer.json'), 'livewire/livewire')) {
            $this->comment('Please, make sure that livewire/livewire package is installed!');
            return 1;
        }
        $tailwind_config_path = './tailwind.config.js';
        if (!file_exists($tailwind_config_path)) {
            $this->comment('Please, make sure that Tailwind CSS is installed and tailwind.config.js file is in the project root folder!');
            return 1;
        } else {
            $tailwind_config = file_get_contents($tailwind_config_path);
            if (!strpos($tailwind_config, 'tall-notifier')) {
                $tailwind_config = preg_replace("#purge:\s*\[\n(.+)\]#imsU", "purge: [\n        \"./config/tall-notifier.php\",\n$1]", $tailwind_config);
                file_put_contents($tailwind_config_path, $tailwind_config);
            }
        }
        $this->call('vendor:publish', ['--tag' => 'tall-notifier.config']);
        $this->call('vendor:publish', ['--tag' => 'tall-notifier.views']);
        $this->info('🥳 Livewire Notifier is installed!');
        return 0;
    }
}


resources/views/livewire/notifier-message.blade.php
  <div x-transition:enter="transition transform-gpu ease-in" x-transition:enter-start="opacity-0 -translate-y-full"
      x-transition:enter-end="opacity-1 translate-y-0" x-transition:leave="transition transform-gpu duration-500"
      x-transition:leave-start="opacity-1" x-transition:leave-end="opacity-0 translate-x-full" x-data="{show: false}"
      x-show="show" 
      @if($message['closable'])@click="show=false" @keydown.escape.window="show=false"@endif x-init="
            setTimeout(()=>{
            show=true
            },50); @if($message['duration'])
            setTimeout(()=>{
                show=false;
                },{{ $message['duration'] }}); @endif">
      <div class="@if($message['closable']) cursor-pointer @endif flex flex-col select-none {{ $message['msgClass'] }} {{ config('tall-notifier.defaultMsgClass')}} overflow-hidden">
          <div class="flex flex-row">
              <div class="flex flex-col items-center justify-center p-2">
                  @if ($message['icon'])
                      {!! $message['icon'] !!}
                  @else
                       @include(config('tall-notifier.types.'.$message['type'].'.icon'))
                  @endif
              </div>
              <div class="flex flex-col items-start justify-center p-2 select-none">
                  @if ($message['text'])
                      <h3 class="text-xl font-bold outline-none">{{ $message['title'] }}</h3>
                  @endif
                  <p class="outline-none">
                  {{ $message['text'] }}
                  </p>
              </div>
          </div>
          @if($message['progressClass'])
          <div x-ref="progress" x-show="show" x-transition:enter-start="translate-x-0"
              x-transition:enter-end="-translate-x-full"
              class="{{ $message['progressClass'] }} h-1 w-full transition transform-gpu animate-pulse"
              style="transition-duration: {{ $message['duration'] }}ms">
          </div>
          @endif
      </div>
  </div>


config/tall-notifier.php
<?php
return [
    'positionClass' => 'absolute top-0 right-0',
    'defaultMsgClass' => 'w-80  rounded-xl shadow-xl',
    'duration' => 5000,
    'types' => [
        'success' => [
            // 'msgClass'=>'bg-green-200',
            'msgClass' => 'bg-gradient-to-r from-green-200 to-green-300',
            'progressClass' => 'bg-green-500',
            'icon' => 'tall-notifier::icons.success',
        ],
        'error' => [
            // 'msgClass'=>'bg-red-200',
            'msgClass' => 'bg-gradient-to-r from-red-200 to-red-300',
            'progressClass' => 'bg-red-500',
            'icon' => 'tall-notifier::icons.error',
        ],
        'fail' => [
            // 'msgClass'=>'bg-red-200',
            'msgClass' => 'bg-gradient-to-r from-red-200 to-red-300',
            'progressClass' => 'bg-red-500',
            'icon' => 'tall-notifier::icons.error',
        ],
        'warning' => [
            // 'msgClass'=>'bg-yellow-200',
            'msgClass' => 'bg-gradient-to-r from-yellow-200 to-yellow-300',
            'progressClass' => 'bg-yellow-500',
            'icon' => 'tall-notifier::icons.error',
        ],
        'warn' => [
            // 'msgClass'=>'bg-yellow-200',
            'msgClass' => 'bg-gradient-to-r from-yellow-200 to-yellow-300',
            'progressClass' => 'bg-yellow-500',
            'icon' => 'tall-notifier::icons.error',
        ],
        'info' => [
            // 'msgClass'=>'bg-blue-200',
            'msgClass' => 'bg-gradient-to-r from-blue-200 to-blue-300',
            'progressClass' => 'bg-blue-500',
            'icon' => 'tall-notifier::icons.info',
        ],
        'default' => [
            // 'msgClass'=>'bg-gray-200',
            'msgClass' => 'bg-gradient-to-r from-gray-200 to-gray-300',
            'progressClass' => 'bg-gray-700',
            'icon' => 'tall-notifier::icons.info',
        ]
    ]
];
